File: backend/app/schemas/audit_calendar.py
"""Strict calendar commands. No client-supplied actor, grant or fact snapshot."""
from datetime import date as Date, datetime
from typing import Annotated, Literal, Union
from uuid import UUID

from pydantic import AfterValidator, BeforeValidator, BaseModel, ConfigDict, Field, StrictBool, StrictInt, StringConstraints, TypeAdapter, model_validator


def calendar_date(value: Date) -> Date:
    if not Date(2000, 1, 1) <= value <= Date(2100, 12, 31):
        raise ValueError("Дата должна быть в диапазоне с 2000 по 2100 год")
    return value


def date_input(value):
    if type(value) is Date:
        return value
    if not isinstance(value, str) or len(value) != 10:
        raise ValueError("Укажите календарную дату в формате ГГГГ-ММ-ДД")
    return value


Day = Annotated[Date, BeforeValidator(date_input), AfterValidator(calendar_date)]
Text = Annotated[str, StringConstraints(strip_whitespace=True, max_length=2000)]
Reason = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1, max_length=2000)]
Code = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1, max_length=40)]
Role = Literal["auditor", "tech", "speaker", "observer"]
Minute = Annotated[StrictInt, Field(ge=0, le=1440, multiple_of=30)]
Count = Annotated[StrictInt, Field(ge=0, le=1000)]


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    @model_validator(mode="before")
    @classmethod
    def literal_types(cls, value):
        if isinstance(value, dict):
            for key in ("confirm", "legacy"):
                if key in value and type(value[key]) is not bool:
                    raise ValueError(f"Поле {key} должно иметь логическое значение")
            if "version" in value and type(value["version"]) is not int:
                raise ValueError("Версия должна быть целым числом")
        return value


class Versioned(StrictModel):
    request_id: UUID
    expected_version: Annotated[StrictInt, Field(ge=0)]


class Setup(StrictModel):
    request_id: UUID
    name: Annotated[str, StringConstraints(strip_whitespace=True, min_length=1, max_length=160)]
    baseline: Day | None = None


class MemberSave(Versioned):
    user_id: UUID
    code: Code
    role: Role
    can_manage: StrictBool
    active: StrictBool


class GroupSave(StrictModel):
    id: UUID | None = None
    code: Code
    label: Annotated[str, StringConstraints(strip_whitespace=True, min_length=1, max_length=160)]
    legacy: Literal[False] = False
    archived: StrictBool = False
    effective_from: Day
    auditor_id: UUID | None = None
    tech_id: UUID | None = None
    reason: Reason


class Meeting(StrictModel):
    date: Day
    start: Minute
    duration: Annotated[StrictInt, Field(ge=30, le=1440, multiple_of=30)]
    activity: Annotated[str, StringConstraints(strip_whitespace=True, max_length=120)]
    speaker_id: UUID | None

    @model_validator(mode="after")
    def interval(self):
        if self.start + self.duration > 1440:
            raise ValueError("Встреча не может переходить через полночь")
        return self


class PlanSave(Meeting):
    id: UUID | None = None
    group_id: UUID
    status: Literal["draft", "planned", "cancelled"]
    reason: Text = ""


class FactRecord(Meeting):
    plan_id: UUID
    group_id: UUID
    outcome: Literal["completed", "cancelled"]
    reason: Reason
    evidence: Reason
    auditor_absent_minutes: Annotated[StrictInt, Field(ge=0, le=1440)]


class PlanRevise(PlanSave):
    id: UUID
    reason: Reason


class Participant(StrictModel):
    user_id: UUID
    role: Role


class FactRestore(Meeting):
    source_row_id: UUID
    outcome: Literal["completed", "cancelled"]
    reason: Reason
    evidence: Reason
    composition_unknown: StrictBool
    participants: Annotated[list[Participant], Field(max_length=30)]
    confirm: Literal[True]

    @model_validator(mode="after")
    def composition(self):
        ids = [p.user_id for p in self.participants]
        if len(ids) != len(set(ids)):
            raise ValueError("Участник повторяется в историческом составе")
        if self.composition_unknown == bool(self.participants):
            raise ValueError("Укажите фактических участников или явно отметьте состав как неизвестный")
        return self


class NoticeRecord(StrictModel):
    plan_id: UUID
    user_id: UUID
    reported_at: datetime
    reason: Reason

    @model_validator(mode="after")
    def aware(self):
        if self.reported_at.utcoffset() is None:
            raise ValueError("Время сообщения должно содержать часовой пояс")
        return self


class AvailabilityPatch(StrictModel):
    date: Day
    start: Minute
    end: Minute
    value: StrictBool | None

    @model_validator(mode="after")
    def interval(self):
        if self.start >= self.end:
            raise ValueError("Конец интервала должен быть позже его начала")
        return self


class AvailabilityPaint(StrictModel):
    user_id: UUID
    patches: Annotated[list[AvailabilityPatch], Field(min_length=1, max_length=672)]

    @model_validator(mode="after")
    def period(self):
        dates = [p.date for p in self.patches]
        if (max(dates) - min(dates)).days > 13:
            raise ValueError("За одну операцию можно изменить доступность не более чем за 14 дней")
        return self


class AbsenceAdd(StrictModel):
    user_id: UUID
    start_date: Day
    end_date: Day
    reason: Reason

    @model_validator(mode="after")
    def period(self):
        if not 0 <= (self.end_date - self.start_date).days < 366:
            raise ValueError("Отсутствие должно длиться от 1 до 366 дней")
        return self


class AbsenceEnd(StrictModel):
    id: UUID
    reason: Reason


class NormSet(StrictModel):
    group_id: UUID | None
    effective_from: Day
    value: Count
    reason: Reason


class ScopeArchive(StrictModel):
    archived: StrictBool
    reason: Reason


class GroupCommand(Versioned):
    operation: Literal["group.save"]
    payload: GroupSave


class PlanCommand(Versioned):
    operation: Literal["plan.save"]
    payload: PlanSave


class FactCommand(Versioned):
    operation: Literal["fact.record"]
    payload: FactRecord


class ReviseCommand(Versioned):
    operation: Literal["plan.revise"]
    payload: PlanRevise


class RestoreCommand(Versioned):
    operation: Literal["fact.restore"]
    payload: FactRestore


class NoticeCommand(Versioned):
    operation: Literal["notice.record"]
    payload: NoticeRecord


class AvailabilityCommand(Versioned):
    operation: Literal["availability.paint"]
    payload: AvailabilityPaint


class AbsenceCommand(Versioned):
    operation: Literal["absence.add"]
    payload: AbsenceAdd


class EndCommand(Versioned):
    operation: Literal["absence.end"]
    payload: AbsenceEnd


class NormCommand(Versioned):
    operation: Literal["norm.set"]
    payload: NormSet


class ArchiveCommand(Versioned):
    operation: Literal["scope.archive"]
    payload: ScopeArchive


Command = Annotated[Union[GroupCommand, PlanCommand, ReviseCommand, FactCommand, RestoreCommand, NoticeCommand, AvailabilityCommand, AbsenceCommand, EndCommand, NormCommand, ArchiveCommand], Field(discriminator="operation")]
command_adapter = TypeAdapter(Command)


class ImportGroupMapping(StrictModel):
    auditor_id: UUID
    tech_id: UUID

    @model_validator(mode="after")
    def distinct(self):
        if self.auditor_id == self.tech_id:
            raise ValueError("Аудитор и техспециалист должны быть разными участниками")
        return self


class ImportPreview(Versioned):
    source: dict
    mapping: Annotated[dict[str, UUID], Field(max_length=1000)]
    bootstrap_history: StrictBool = False
    group_mapping: Annotated[dict[str, ImportGroupMapping], Field(max_length=1000)] = Field(default_factory=dict)


class ImportApply(Versioned):
    confirm: Literal[True]
    reason: Reason

File: frontend/src/api/auditCalendar.ts
import { api } from './client'

export type CalendarRole = 'auditor' | 'tech' | 'speaker' | 'observer'
export type CalendarScope = { id: string; name: string; timezone: string; baseline: string; version: number; today: string; now: string; archived: boolean; history_complete: boolean }
export type CalendarMember = { user_id: string; full_name: string; code: string; role: CalendarRole; can_manage: boolean; active: boolean }
export type CalendarGroup = { id: string; code: string; label: string; legacy: boolean; archived: boolean; versions: { id: string; effective_from: string; auditor_id: string | null; tech_id: string | null }[] }
export type CalendarIssue = { code: string; message: string; user_id?: string; participant_code?: string; participant_name?: string; role?: CalendarRole; record_id?: string }
export type CalendarMeetingOptions = { version: number; date: string; start: number; duration: number; groups: { id: string; code: string; label: string; eligible: boolean; issues: CalendarIssue[]; warnings: CalendarIssue[] }[] }
export type CalendarMeetingWindowCell = { date: string; start: number; status: 'available' | 'warning' | 'unavailable' | 'expired'; confirmed: number; uncertain: number }
export type CalendarMeetingWindowQuery = { date: string; start: number; duration: number; group_id: string | null; speaker_id: string | null }
export type CalendarMeetingWindowOption = { group_id: string; group_version_id: string; auditor_id: string; tech_id: string; speaker_id: string; status: 'available' | 'warning'; warnings: CalendarIssue[] }
export type CalendarMeetingWindows = {
  version: number; now: string
  period: { from: string; to: string; duration: number; group_id: string | null; speaker_id: string | null; full_day: boolean }
  cells: CalendarMeetingWindowCell[]
}
export type CalendarMeetingWindowOptions = { version: number; now: string; query: CalendarMeetingWindowQuery; options: CalendarMeetingWindowOption[] }
export type CalendarMeetingWindowPrefill = { duration: number; group_id: string; speaker_id: string; version: number; server_now?: string; clock_started?: { wall: number; monotonic: number } }
export type CalendarWorkload = {
  version: number; period: { from: string; to: string; group_id: string | null }; working_days: number
  working_window: { start: number; end: number; slot_minutes: number }
  members: (CalendarMember & { filled_days: number; partial_days: number; unfilled_days: number; absence_days: number
    free_slots: number; free_minutes: number; planned_meetings: number; planned_minutes: number; planned_work_minutes: number
    outside_work_minutes: number; target: number; power_percent: number | null; norm_percent: number | null })[]
}
export type CalendarPlan = { id: string; date: string; start: number; duration: number; group_id: string; group_version_id: string | null; activity: string; speaker_id: string | null; status: 'draft' | 'planned' | 'cancelled'; fact_outcome?: 'completed' | 'cancelled' | null; version: number; origin: string; source_id: string | null; issues: CalendarIssue[]; warnings: CalendarIssue[] }
export type CalendarFact = { id: string; plan_id: string | null; date: string; start: number; duration: number; group_id: string | null; activity: string; speaker_id: string | null; outcome: 'completed' | 'cancelled'; reason: string; evidence: string; recorded_by_id: string; recorded_at: string; participant_snapshot: unknown[]; planned_snapshot: unknown; composition_unknown: boolean; origin: string }
export type CalendarAvailability = { user_id: string; date: string; start: number; end: number; available: boolean }
export type CalendarAvailabilityLock = { id: string; user_id: string; date: string; locked: boolean; locked_at: string; locked_by_id: string; snapshot: unknown[] }
export type CalendarChangeRequest = { id: string; user_id: string; date: string; reason: string; status: 'pending' | 'approved' | 'closed' | 'rejected'; requested_at: string; requested_by_id: string; opened_at: string | null; opened_by_id: string | null; closed_at: string | null; closed_by_id: string | null; resolution: string; before: unknown[]; after: unknown[] | null }
export type CalendarReadiness = {
  scope_version: number; from: string; to: string; duration: number; working_start: number; working_end: number
  employees: { user_id: string; full_name: string; code: string; role: CalendarRole; days: { date: string; status: 'missing' | 'partial' | 'filled' | 'absent'; free_minutes: number; locked: boolean }[]; filled_days: number; total_days: number }[]
  groups: { group_id: string; code: string; label: string; days: { date: string; member_ids: string[]; missing_user_ids: string[]; status: 'no_composition' | 'missing' | 'absent' | 'no_overlap' | 'booked' | 'available'; common_windows: { start: number; end: number }[]; free_windows: { start: number; end: number }[]; slots: { start: number; end: number }[]; plans: { id: string; start: number; duration: number; activity: string; status: string }[]; last_notified_at: string | null }[] }[]
}
export type CalendarAbsence = { id: string; user_id: string; start_date: string; end_date: string; reason: string; version: number; status: 'active' | 'cancelled' }
export type CalendarTimelineMeeting = { id: string; kind: 'plan' | 'fact'; start: number; duration: number; activity: string; status: string; group_code: string; participants: { user_id: string; role: CalendarRole }[] }
export type CalendarTimeline = { version: number; date: string; members: CalendarMember[]; availability: CalendarAvailability[]; absences: CalendarAbsence[]; locks: CalendarAvailabilityLock[]; meetings: CalendarTimelineMeeting[] }
export type CalendarNorm = { id: string; group_id: string | null; effective_from: string; value: number; reason: string; recorded_at: string; recorded_by_id: string }
export type CalendarNotice = { id: string; plan_id: string; user_id: string; reported_at: string; reason: string }
export type CalendarTotals = { target: number; completed: number; balance: number; backlog: number }
export type CalendarStats = { plan: number; fact: number; attention: number; target: number; backlog: number; through: string | null; target_scope: 'team' | 'group'; groups: ({ group_id: string; code: string } & CalendarTotals)[]; fortnights: ({ from: string; to: string; cumulative_backlog: number } & CalendarTotals)[] }
export type CalendarState = { scope: CalendarScope; actor: { user_id: string; can_manage: boolean; can_archive: boolean }; members: CalendarMember[]; groups: CalendarGroup[]; plans: CalendarPlan[]; facts: CalendarFact[]; availability: CalendarAvailability[]; availability_locks: CalendarAvailabilityLock[]; change_requests: CalendarChangeRequest[]; absences: CalendarAbsence[]; norms: CalendarNorm[]; notices: CalendarNotice[]; stats: CalendarStats }
export type AvailabilityPatch = { date: string; start: number; end: number; value: boolean | null }
export type CalendarCommandMap = {
  'group.save': { id?: string; code: string; label: string; legacy?: false; archived?: boolean; effective_from: string; auditor_id?: string; tech_id?: string; reason: string }
  'plan.save': { id?: string; date: string; start: number; duration: number; group_id: string; activity: string; speaker_id: string | null; status: CalendarPlan['status']; reason?: string }
  'plan.revise': { id: string; date: string; start: number; duration: number; group_id: string; activity: string; speaker_id: string | null; status: CalendarPlan['status']; reason: string }
  'fact.record': { plan_id: string; date: string; start: number; duration: number; group_id: string; activity: string; speaker_id: string | null; outcome: CalendarFact['outcome']; reason: string; evidence: string; auditor_absent_minutes: number }
  'fact.restore': { source_row_id: string; date: string; start: number; duration: number; activity: string; speaker_id: string | null; outcome: CalendarFact['outcome']; reason: string; evidence: string; composition_unknown: boolean; participants: { user_id: string; role: CalendarRole }[]; confirm: true }
  'notice.record': { plan_id: string; user_id: string; reported_at: string; reason: string }
  'availability.paint': { user_id: string; patches: AvailabilityPatch[]; expected?: AvailabilityPatch[] }
  'availability.lock': { user_id: string; date: string; reason: string }
  'availability.request': { user_id: string; date: string; reason: string }
  'availability.resolve': { id: string; action: 'approve' | 'close' | 'reject'; reason: string }
  'availability.notify': { group_id: string; date: string; reason: string; duration?: number }
  'absence.add': { user_id: string; start_date: string; end_date: string; reason: string }
  'absence.end': { id: string; reason: string }
  'norm.set': { group_id: string | null; effective_from: string; value: number; reason: string }
  'scope.archive': { archived: boolean; reason: string }
}
export type CalendarCommand = { [K in keyof CalendarCommandMap]: { operation: K; payload: CalendarCommandMap[K] } }[keyof CalendarCommandMap]
export type CalendarMutation = { version: number; result: unknown }
export type CalendarAdminState = { scope: CalendarScope | null; members: CalendarMember[]; users: { id: string; full_name: string; email: string; audit_calendar_enabled: boolean; is_active: boolean }[] }
export type CalendarHistory = { items: { id: string; action: string; actor_name: string; occurred_at: string; detail: unknown }[]; total: number }
export type CalendarImport = { id: string; version: number; status: string; summary: unknown; issues: unknown[]; rows: unknown[]; mapping_required: unknown[]; source_sha256: string }
const root = '/api/audit-calendar'
export const auditCalendar = {
  state: (params: Record<string, string>, signal?: AbortSignal) => api.get<CalendarState>(`${root}/state`, params, { signal }),
  readiness: (params: { from: string; to: string; duration: number }, signal?: AbortSignal) => api.get<CalendarReadiness>(`${root}/readiness`, { ...params, duration: String(params.duration) }, { signal }),
  timeline: (date: string, signal?: AbortSignal) => api.get<CalendarTimeline>(`${root}/availability-timeline`, { date }, { signal }),
  meetingOptions: (params: { date: string; start: number; duration: number; speaker_id?: string; plan_id?: string }, signal?: AbortSignal) => api.get<CalendarMeetingOptions>(`${root}/meeting-options`, { ...params, start: String(params.start), duration: String(params.duration) }, { signal }),
  meetingWindows: (params: { from: string; to: string; duration: number; group?: string; speaker_id?: string; full_day: boolean }, signal?: AbortSignal) => api.get<CalendarMeetingWindows>(`${root}/meeting-windows`, { ...params, duration: String(params.duration), full_day: String(params.full_day) }, { signal }),
  meetingWindowOptions: (params: { date: string; start: number; duration: number; group?: string; speaker_id?: string }, signal?: AbortSignal) => api.get<CalendarMeetingWindowOptions>(`${root}/meeting-window-options`, { ...params, start: String(params.start), duration: String(params.duration) }, { signal }),
  workload: (params: { from: string; to: string; group?: string }, signal?: AbortSignal) => api.get<CalendarWorkload>(`${root}/workload`, params, { signal }),
  command: (command: CalendarCommand, request_id: string, expected_version: number) => api.post<CalendarMutation>(`${root}/commands`, { request_id, expected_version, ...command }),
  admin: () => api.get<CalendarAdminState>(`${root}/admin`),
  setup: (body: { request_id: string; name: string; baseline: string }) => api.post<CalendarMutation>(`${root}/admin/setup`, body),
  member: (body: { request_id: string; expected_version: number; user_id: string; code: string; role: CalendarRole; can_manage: boolean; active: boolean }) => api.post<CalendarMutation>(`${root}/admin/members`, body),
  history: (limit = 100) => api.get<CalendarHistory>(`${root}/history`, { limit: String(limit) }),
  imports: () => api.get<CalendarImport[] | { items: CalendarImport[] }>(`${root}/imports`),
  preview: (body: { request_id: string; expected_version: number; source: unknown; mapping: Record<string, string>; bootstrap_history?: boolean; group_mapping?: Record<string, { auditor_id: string; tech_id: string }> }) => api.post<CalendarImport>(`${root}/imports/preview`, body),
  applyImport: (id: string, body: { request_id: string; expected_version: number; confirm: true; reason: string }) => api.post<CalendarMutation>(`${root}/imports/${encodeURIComponent(id)}/apply`, body),
}
